<launch>

  <arg name="robot" default="cob4-8" />
  <!-- cob4-5, cob4-7, cob4-8, cob4-10, cob4-13, cob4-16, cob4-24 -->
  <arg name="robot_env" default="empty" />
  <!-- empty, ipa-apartment, ipa-kitchen, ipa-office -->
  <arg name="pkg_hardware_config" default="$(find cob_hardware_config)"/>
  <arg name="pkg_moveit_config" default="$(find cob_moveit_config)"/>
  <arg name="debug" default="false"/>
  <arg name="pipeline" default="ompl" />
  <arg name="scene_objects" default="pick_place" />

  <arg name="moveit" default="false"/>
  <arg name="mapping" default="false" />
  <arg name="localization" default="false" />
  <arg name="navigation" default="false" />

  <arg name="world_name" default="$(find pedsim_simulator)/worlds/empty_default.world" />
  <!-- test_zone.sdf, test_zone_pick_place.sdf -->
  <arg name="output" default="log"/>
  <arg name="allow_active_sensing" default="true"/>
  <arg name="vicon_bridge_params_file" default="localization_tests.yaml"/>
  <arg name="prepare_robot_pick_place" default="true" />
  <arg name="gui" default="true"/>

  <!-- initial position of robot for localization testing -->
  <arg name="x_pose" default="0.0"/>
  <arg name="y_pose" default="0.0"/>

  <!--pedsim parameters config-->
  <arg name="kbd_teleop" default="false"/>
  <arg name="rqt_teleop" default="false"/>
  <arg name="visualize" default="true"/>
  <arg name="with_robot" default="true"/>

  <arg name="simulation_factor" default="1.0"/>
  <!-- Speed up -->
  <arg name="update_rate" default="25.0"/>
  <!-- Hz -->

  <!-- additional parameters configured -->
  <arg name="force_obstacle" default="10.0"/>
  <arg name="sigma_obstacle" default="0.2"/>
  <arg name="force_social" default="5.1"/>
  <arg name="force_group_gaze" default="3.0"/>
  <arg name="force_group_coherence" default="2.0"/>
  <arg name="force_group_repulsion" default="1.0"/>
  <arg name="force_random" default="0.1"/>
  <arg name="force_wall" default="2.0"/>

  <arg name="scenario_file" default="ipa_apartment.xml"/>
  <arg name="detect_frozen_agents" default="false"/>
  <arg name="walls_scale" default="0.1"/>

  <!--additional parameters for cob simulation-->
  <arg name="frame_id" default="map"/>
  <arg name="robot_base_frame_id" default="base_footprint"/>
  <arg name="obstacle_offset_x" default="-19.2"/>
  <arg name="world_frame" default="map"/>
  <arg name="robot_frame" default="base_footprint"/>
  <arg name="publish_tf" default="false"/>


  <!--end of args and parameters-->

  <!--pedsim simulator-->
  <!-- Simulator -->
  <include file="$(find pedsim_simulator)/launch/simulator.launch">
    <arg name="kbd_teleop" value="$(arg kbd_teleop)"/>
    <arg name="rqt_teleop" value="$(arg rqt_teleop)"/>
    <arg name="scene_file" value="$(find pedsim_simulator)/scenarios/$(arg scenario_file)"/>
    <arg name="with_robot" value="$(arg with_robot)"/>
    <arg name="simulation_factor" value="$(arg simulation_factor)"/>
    <arg name="update_rate" value="$(arg update_rate)"/>
    <arg name="default_queue_size" value="10"/>
    <arg name="max_robot_speed" value="1.5"/>
    <arg name="robot_mode" value="1"/>
    <arg name="enable_groups" value="true"/>
    <arg name="pose_initial_x" value="10.0"/>
    <arg name="pose_initial_y" value="15.0"/>
    <arg name="pose_initial_theta" value="0.7"/>
    <arg name="detect_frozen_agents" value="$(arg detect_frozen_agents)" />

    <!-- extra parameters -->
    <arg name="force_obstacle" value="$(arg force_obstacle)"/>
    <arg name="sigma_obstacle" value="$(arg sigma_obstacle)"/>
    <arg name="force_social" value="$(arg force_social)"/>
    <arg name="force_group_gaze" value="$(arg force_group_gaze)"/>
    <arg name="force_group_coherence" value="$(arg force_group_coherence)"/>
    <arg name="force_group_repulsion" value="$(arg force_group_repulsion)"/>
    <arg name="force_random" value="$(arg force_random)"/>
    <arg name="force_wall" value="$(arg force_wall)"/>

    <!--additional parameters for cob simulation-->
    <arg name="frame_id" value="$(arg frame_id)"/>
    <arg name="robot_base_frame_id" value="$(arg robot_base_frame_id)"/>
    <arg name="obstacle_offset_x" value="$(arg obstacle_offset_x)"/>
    <arg name="world_frame" value="$(arg world_frame)"/>
    <arg name="robot_frame" value="$(arg robot_frame)"/>
    <arg name="publish_tf" value="$(arg publish_tf)"/>
  </include>

  <!-- Visualizer -->
  <include file="$(find pedsim_simulator)/launch/visualizer.launch">
    <arg name="walls_scale" value="$(arg walls_scale)"/>
  </include>

  <!-- Rviz -->
  <node pkg="rviz" type="rviz" name="rviz" args="-d $(find pedsim_simulator)/rviz/cob_ped_sim.rviz" if="$(arg visualize)"/>


  <!--COB Gazebo simulation-->
  <!-- Start Gazebo simulation -->
  <include file="$(find cob_bringup_sim)/launch/robot.launch">
    <arg name="robot" value="$(arg robot)" />
    <arg name="robot_env" value="$(arg robot_env)" />
    <arg name="gui" value="$(arg gui)" />
    <arg name="initial_config" value="-x $(arg x_pose) -y $(arg y_pose)" />
    <arg name="world_name" value="$(arg world_name)"/>
  </include>

  <!-- MoveIt! -->
  <group if="$(arg moveit)">
    <include file="$(find cob_moveit_bringup)/launch/move_group.launch">
      <arg name="robot" value="$(arg robot)" />
      <arg name="pkg_hardware_config" value="$(arg pkg_hardware_config)"/>
      <arg name="pkg_moveit_config" value="$(arg pkg_moveit_config)"/>
      <arg name="load_robot_description" value="true"/>
      <arg name="allow_trajectory_execution" value="true"/>
      <arg name="fake_execution" value="false"/>
      <arg name="use_sensors" value="false"/>
      <arg name="info" value="true"/>
      <arg name="debug" value="$(arg debug)"/>
      <arg name="pipeline" value="$(arg pipeline)" />
    </include>
  </group>

  <!-- Moving COB4 to its home state -->
  <node pkg="test_cob4_common_bringup" type="cob4_moveit_home" name="cob4_moveit_home" output="screen" />

  <!-- gazebo physics parameters changed for faster simulation -->
  <node name="gazebo_physics_config" pkg="dynamic_reconfigure" type="dynparam" args="set_from_parameters gazebo">
    <rosparam file="$(find test_cob4_simulation_bringup)/param/physics_gazebo.yaml" command="load" />
  </node>

  <!-- Mapping! -->
  <group if="$(arg mapping)">
    <include file="$(find cob_mapping_slam)/launch/2dslam.xml"/>
  </group>

  <!-- Localization is only possible when Mapping is set to false -->
  <group unless="$(arg mapping)">
    <group if="$(arg localization)">

      <!-- pre-made map for localization -->
      <arg name="map_file" default="$(find test_cob4_simulation_bringup)/maps/map.yaml"/>
      <!-- Map server -->
      <node pkg="map_server" name="map_server" type="map_server" args="$(arg map_file)"/>

      <!-- start amcl -->
      <node pkg="amcl" type="amcl" name="amcl">
        <remap from="scan" to="scan_unified" />
        <param name="odom_model_type" value="omni"/>
        <param name="transform_tolerance" value="0.2" />
        <param name="gui_publish_rate" value="-1.0"/>
        <param name="laser_max_beams" value="180"/>
        <param name="laser_max_range" value="10"/>
        <param name="min_particles" value="100"/>
        <param name="max_particles" value="5000"/>
        <!--param name="update_min_a" value="1.0"/-->
        <param name="kld_err" value="0.01"/>
        <param name="kld_z" value="0.99"/>
        <param name="odom_alpha1" value="0.1"/>
        <param name="odom_alpha2" value="0.1"/>
        <!-- translation std dev, m -->
        <param name="odom_alpha3" value="0.1"/>
        <param name="odom_alpha4" value="0.1"/>
        <param name="odom_alpha5" value="0.1"/>
        <param name="laser_z_hit" value="0.95"/>
        <param name="laser_z_short" value="0.1"/>
        <param name="laser_z_max" value="0.05"/>
        <param name="laser_z_rand" value="0.05"/>
        <param name="laser_sigma_hit" value="0.2"/>
        <param name="laser_lambda_short" value="0.1"/>
        <param name="laser_model_type" value="likelihood_field"/>
        <!-- <param name="laser_model_type" value="beam"/> -->
        <param name="laser_likelihood_max_dist" value="2.0"/>
        <param name="update_min_d" value="0.2"/>
        <param name="update_min_a" value="0.2"/>
        <param name="odom_frame_id" value="odom_combined"/>
        <param name="base_frame_id" value="base_footprint"/>
        <param name="global_frame_id" value="map"/>
        <param name="resample_interval" value="2"/>
        <param name="transform_tolerance" value="0.1"/>
        <param name="recovery_alpha_slow" value="0.0"/>
        <param name="recovery_alpha_fast" value="0.0"/>
      </node>
    </group>
  </group>

  <group unless="$(arg mapping)">
    <group if="$(arg localization)">
      <group if="$(arg navigation)">
        <arg name="pkg_nav_config" default="$(find cob_navigation_config)"/>

        <!--- Run move base -->
        <include file="$(find cob_navigation_global)/launch/2dnav_ros_dwa.xml">
          <arg name="robot" value="$(arg robot)"/>
          <arg name="pkg_nav_config" value="$(arg pkg_nav_config)"/>
        </include>
      </group>
    </group>
  </group>

  <!-- this node spawn pedsim actors to gazebo once, then the plugin updates their pose -->
  <node pkg="pedsim_gazebo_plugin" type="spawn_pedsim_agents.py" name="spawn_pedsim_agents" output="screen">
  </node>

</launch>
